src/libsystemd/sd-bus/bus-error.c

```diff
@@ -536,7 +536,6 @@ _public_ int sd_bus_error_set_errnofv(sd_bus_error *e, int error, const char *fo
 
                                 free(m);
                         } else {
-                                free((char*) e->message);
```

Comment:
`e->message` is heap allocated when `e->_need_free` is > 0, which is checked beforehand here, hence the pre-existing code is correct.

Reply:
Hi @poettering  , I have below query (for understanding purpose only) in bus-error.c file.

In function, 
_**public_ int sd_bus_error_set_errnofv(sd_bus_error *e, int error, const char *format, va_list ap) {**
...
...  // some code
...
 ***e = errno_to_bus_error_const(error);**    _//  dbus-error.c:500 Here, e->message will be pointing to some string literal (e.g. "Inconsistent message") which is not heap allocated._
...
... // some code
...
        if (format) {  // if condition evaluates to true
....
.... // some code
....
        if (r >= 0) {   // if condition evaluates to true
            if (e->_need_free <= 0) {  // if condition evaluates to false
...
... // some code
...
            } else { 
                **free((char*) e->message);**    _// dbus-error.c:539 free is called on e-> message which is not heap allocated,
                                                             // since e->message is still pointing to string literal
                                                             //  ( initialized to string literal at line dbus-error.c:500 above)_
                                                                                                                  
...
...
            } // end of else part
        } // end of  if (r >= 0) 
    } // end of if ( format) 
**bus_error_strerror(e, error);**  // _dbus-error.c:547.  e->message is heap allocated here._
...
} // end of function

Could not find any call to heap allocate memory pointed to by e->message after dbus-error.c:500 till line number dbus-error.c:547
free is called on e->message at line dbus-error.c:537. 

Could you kindly explain this flow.

Reply:
errno_to_bus_error_const() sets `_need_free` to zero if the message is a static, literal string. It's not heap allocated.

Your static analyzer is too simple here: the message in some cases is a static literal string and in other cases allocated on the heap, and `_need_free` indicates what the case is. But your analyzer cannot deal with thta, it assumes that it's always heap, or always literal, but doesn't understand the case where it can be either, and where we remember which it is so that we can destroy it properl

Reply:
I think @amitanandd's analysis is correct. `e->message` is constant string here. BTW, I've created #16854.